developer/src/kmc-convert/src/keylayout-to-kmn/keylayout-to-kmn-converter.ts

```diff
@@ -1,27 +1,990 @@
+
+  private options: CompilerOptions;
 
-  constructor(/*private*/ _callbacks: CompilerCallbacks, /*private*/ _options: CompilerOptions) {
-    // TODO: if these are needed, uncomment /*private*/ and remove _, and they will then
-    // be available as class properties
+  async init(callbacks: CompilerCallbacks, options: CompilerOptions): Promise<boolean> {
+    this.options = { ...options };
+    this.callbacks = callbacks;
+    return true;
   }
 
-  async run(inputFilename: string, outputFilename: string, binaryData: Uint8Array): Promise<ConverterToKmnArtifacts> {
-    if(!inputFilename || !outputFilename || !binaryData) {
+  constructor(private callbacks: CompilerCallbacks, options: CompilerOptions) { };
+
+  /**
+   * @brief  member function to run read/convert/write
+   * @param  inputFilename the ukelele .keylayout-file to be converted
+   * @param  outputFilename the resulting keyman .kmn-file
+   * @return null on success
+   */
+  async run(inputFilename: string, outputFilename: string = ""): Promise<ConverterToKmnArtifacts> {
+
+    if (!inputFilename) {
       throw new Error('Invalid parameters');
     }
 
-    console.error('TODO: implement KeylayoutToKmnConverter');
+    const KeylayoutReader = new KeylayoutFileReader(this.callbacks/*, this.options*/);
+    const jsonO: object = KeylayoutReader.read(inputFilename);
+    if (!jsonO) {
+      throw new Error('Error while processing read()');
+    }
+
+    const outArray: convert_object = await this.convert(jsonO, outputFilename);
+    if (!outArray) {
+      throw new Error('Error while processing convert()');
+    }
+
+    const kmnFileWriter = new KmnFileWriter(this.callbacks, this.options);
+    const out_text_ok: boolean = kmnFileWriter.write(outArray);
 
+    if (!out_text_ok) {
+      throw new Error('Error while processing write()');
+    }
     return null;
   }
-}
\ No newline at end of file
+
+  /**
+   * @brief  member function to read filename and behaviour of a json object into a convert_object
+   * @param  jsonObj containing filename, behaviour and rules of a json object
+   * @return an convert_object containing all data ready to print out
+   */
+  public convert(jsonObj: any, outputfilename: string = ""): convert_object {
+
+    const modifierBehavior: string[][] = [];           // modifier for each behaviour
+    const rules: Rule[] = [];                          // an array of data for a kmn rule
+
+    const data_object: convert_object = {
+      keylayout_filename: "",
+      kmn_filename: "",
+      arrayOf_Modifiers: [],
+      arrayOf_Rules: []
+    };
+
+    // ToDo in a new PR: check tags  ( issue # 13599)
+    if ((jsonObj !== null) && (jsonObj.hasOwnProperty("keyboard"))) {
+
+      // get filename from data that had been read
+      const fileNameNoExtention = jsonObj.keyboard['@_name'];
+      const filePath = this.callbacks.path.join(process.cwd(), KeylayoutToKmnConverter.TEST_SUBFOLDER, KeylayoutToKmnConverter.DATA_SUBFOLDER);
```

Comment:
We can't call `process.cwd()` here -- do we really need the input filename in this part of the process? We have the input filename already passed into the `run()` function, so if we need it, we should pass it in.